### DISCUSSION TITLE
Supress one incoming alarm

### DISCUSSION THREAD

**Discussion by KaeroDot** [astlouys/Pico-Green-Clock#24] [Ideas]
Sometimes I wake up few minutes before the alarm time. I want to leave the room but I know the alarm will start in few minutes and if the alarm is not snoozed early it will potentially wake up whole family. So either I have to wait till the alarm time or I have to return to the room in few minutes and stop the alarm.
I have seen an interesting feature on some android phones - suppress one incoming alarm.
So I imagine it should work in this way. If I long press bottom button (in normal clock mode), a text will scroll, e.g. "Disabling next alarm". In the firmware one flag, e.g. DisableIncomingAlarm, will be switched from 0 to 1. Later, when any alarm should start, the function checking alarms first looks if DisableIncomingAlarm is 1. If so, the alarm is considered as finished and the flag is set to 0. In this way next alarms will work as normally.

I am not a c coder, but I can add a new flag. I have found the Alarm is started on line 17242 (correct?). So I suppose if somewhere here I add a line that checks the DisableIncomingAlarm, and I will not set the flag AlarmReachedBitMask, the alarm will not start, true?
The DisableIncomingAlarm should be set on line 7926, I believe.

Let me know what you think about this function.

**Comment by astlouys** [comment]
Hi KaeroDot,

Yes, this is a good idea and this is a function I was planning to implement after reading the Green Clock product review in the "Hackable" Magazine No.44 (Sep-Oct-2022).

The review was based on the original Firmware Version (1.00). Many of the problems (or missing features) reported in the review were already added in the new Firmware at that time, but "shutting down in advance an alarm to come" was one of the interesting features I thought could be added... There is already a "silent period" available with the remote control, so I thought the new feature could use this feature already in place and simply improve it...

Line 17242 is indeed the one setting the flag for alarm ringer. I think some timing needs to be considered in the algorithm. It should not be allowed to "press a button" that could disable by mistake an alarm that is supposed to sound in 5 hours from now (or should it ?). Since there is a text associated with each alarm (which is modifiable by user, I guess this is the text you should scroll when saying "Disabling alarm No. xxx - text of the alarm". 

The timer callback (within which is line 17242) is executed only once a second, so the flag you want to implement could be reset as soon as it is processed to prevent the corresponding bit mask to be set.

**Reply by KaeroDot** [reply]
In my previous comment I have incorrectly identified which part of code reacts to button events. Now I got it working. See diff of commit f48be39651daa779cf4fbd914c26ec525149572f in my branch https://github.com/KaeroDot/Pico-Green-Clock/commits/supress_alarm

Ad timing for disabling: it is sound, yet I cannot present any real reasons for/against any time limit. My feeling is one should be able to disable alarm at least one hour before the alarm time. My real use case is disabling max. 30 minutes before the alarm time.

Yes, there should be a text which alarm will be disabled. Yet finding which alarm is the next one is out of my knowledge of the code.

And I have found it could be also useful to have a text scrolling at the time of the disabled alarm, something like "The Alarm 1 was disabled."

With the czech language and ability to disable an incoming alarm, the clock now got every ability I hoped for bedroom clock, so many thanks for your code and help. 👍

**Reply by astlouys** [reply]
I also appreciate this product. As I said on another thread, too bad that there are not more similar "end-user" products base on the Pico / PicoW...

The idea of adding some kind of "timing" for "alarm disable" is simply to prevent disabling an alarm by mistake. For example, during the evening, you press a wrong button, and instead of triggering a legitimate function, you disable "next alarm" which is tomorrow morning at 8h00... By adding some timing, I meant that you cancel nothing if next alarm is farther than - for example - 1 hour away in the future. That being said, it may not be useful (or apply) in your case and you can definitely live without it !.

Also, if you want to scroll the text for the target alarm, it is stored in the variable:  FlashConfig.Alarm[Loop1UInt8].Text
(where Loop1UInt8 corresponds to the alarm number).

As I said, I keep your ideas for future releases ;-)
